src/LearnMe.Core/Services/Calendar/GoogleCalendar.cs

```diff
@@ -0,0 +1,159 @@
+﻿using System;
+using LearnMe.Core.DTO.Calendar;
+using LearnMe.Core.Interfaces.Services;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using AutoMapper;
+using Google.Apis.Calendar.v3;
+using Google.Apis.Calendar.v3.Data;
+using LearnMe.Core.Services.Calendar.Utils;
+using Microsoft.Extensions.Logging;
+using LearnMe.Infrastructure.Models.Domains.Calendar;
+using LearnMe.Infrastructure.Repository.Interfaces;
+
+namespace LearnMe.Core.Services.Calendar
+{
+    public class GoogleCalendar : ICalendar
+    {
+        private readonly CalendarService _calendarService;
+        private readonly string _applicationName = "Learn Me WEB Applicaton";
+        private readonly ICrudRepository<CalendarEvent> _repository;
+        private readonly IMapper _mapper;
+        private readonly ILogger<GoogleCalendar> _logger;
+
+        public GoogleCalendar(IGoogleAPIconnection googleAPIconnection,
+                              ICrudRepository<CalendarEvent> repository,
+                              IMapper mapper,
+                              ILogger<GoogleCalendar> logger)
+        {
+            var token = googleAPIconnection.GetToken();
```

Comment:
Metoda `GetToken()` zwraca jakiś token który bierzecie z `Task<UserCrendetial>` (czy coś). Wywoływanie kodu asynchronicznego w konstruktorze jest na maxa złe. 

W metodzie `GetToken()` macie wywołanie metody `Task.Result`. Zacznijmy od tego, że zamist normalnego wyjątku dostaniecie z niego `AggregateException`, o ile w ogóle coś dostaniecie. 

Jak chcecie coś takiego robić, zamieńcie w implementacji metody `GetToken()` `Result` na `...GetAwaiter().GetResult()`, nie dostajecie wtedy aggregate tylko "normalny" wyjątek (nie chce mi się opisywać różnicy, znajdziesz na necie).

Kolejna sprawa, wywołując kod asynchroniczncy w sposób synchroniczny blokujecie cały wątek. Google to nie jest wasz serwis, może być tak, że on NIGDY nie zadziała bo np. serwery google nie odpowiadają. Wtedy możecie czekać na `.Result` w nieskończoność, bo nie wykona się task (a Result czeka do zakończenia taska). Cały thread na CPU zatrzyma się, bo czeka na wykonanie zadania. Usuń te metodę z konstruktora i przenieś ją do jakiejś metody. Konstruktor ma pozostać synchroniczny i nie wykonywać żadnej logiki. Tylko przypisanie do pól.

Reply:
Poprawiłam pierwszą część - tzn. na awaita, żeby nie mieć AggregateException tylko 1szy wyjątek jakby co.

Jeśli chodzi o drugą część - na razie nie mam pomysłu, gdzie zaincjować ten serwis. Czy ja w takim razie muszę jakoś klasę CalendarService (google'owa) w jakiś sposób rozszerzyć do jakiejś mojej klasy z interfejsem i ją zarejestrować w startupie? tylko nadal ta jedna metoda zwraca mi serwis = tzn. połączenie z kalendarzem i nie wiem jeszcze, gdzie to można włożyć oprócz knstruktora tak szczerze mówiąc. Jedyny pomysł to na początku każdej metody if(serwis nie istnieje) { to wywołaj metodę tworzącą serwis i zapisz ją do pola prywatnego }

Reply:
Ogólnie żadnych metod asynchronicznych nie powinno być w konstruktorze. GetAwaiter.GetResult jest lepsze od Result ale no, i tak to powinno być wyrzucone poza konstruktor.

Jeśli macie logikę tokena w ten sposób zrobioną, że jest pobierany za każdym razem na żądanie to dobrze by było wstrzyknąć go do IoC i wołać ten token w konstruktorze (tak jak normalne serwisy). 

Przykład poniżej z mojego projekut. Ten kod wstrzykuje do IoC informacje o aktualnym użytkowniku który uruchamia requesty:

```csharp
            services.AddTransient(x =>
            {
                var context = x.GetService<IHttpContextAccessor>().HttpContext;
                context.Items.TryGetValue(Const.CURRENT_USER, out var currentUser);

                return (CurrentUser)currentUser
                       ?? throw new ArgumentNullException(nameof(currentUser));
            });
```

I w kodzie wywołuje go w ten sposób:

```csharp

private readonly CurrentUser _currentUser;

public Hub(CurrentUser currentUser)
{
      _currentUser= currentUser
                ?? throw new ArgumentNullException(nameof(currentUser));
            
}
````

Tak samo możesz zarejestrować swój token który będzie dostępny z każdego miejsca w twoim api i będzie pobierany tylko raz.

Co do unit of worka, jeśli calendar service nie jest twoim kode a googlowskim to sorry, mój błąd, nie doczytałem. Wtedy od biedy może to być tak zrobione ale i tak opakowałbym to w ajkiś interfejs który też wstrzykniesz w startupie

Reply:
Tego się nie da zrobić, bo jak opakowuję Googlowy CalendarService w IService to tracę dostęp do jego propercji itp bym musiała przepisywać na nowo metody IService, żeby sie dobierały do tych danych........

Reply:
przyklad: 
![obraz](https://user-images.githubusercontent.com/19174792/94424863-31a65b80-018b-11eb-86a3-15314b7486d1.png)

Reply:
@AgnieszkaAnd  interfejs to kontrakt, ty się umawiasz z kodem, co powinien robić. 

Jeśli `ICalendarService` jest twoim nowo utworzonym interfejsem to lepiej z punktu widzenia OOP byłoby utworzyć nową metodę `InsertEventAsync()` zamiast wywoływać propertę `Events` a potem metodę `Insert`. Dzięki temu będziesz mogła w przyszłości podmienić implementacje do  serwisu kalendarza i nie uzależniasz się od API Google.

Mając interfejs `ICalendarService` z metodami jak `InsertEventAsync()` nie uzależniasz się np. od Google tylko raz możesz użyć Outlooka, raz czegoś innego podmieniając implementację interfejsu. Dzięki temu podmieniając implementacje nie rozwalasz kodu, bo bazujesz na interfejsie który nic nie wie o implementacji

Reply:
![obraz](https://user-images.githubusercontent.com/19174792/94583052-4b27d000-027d-11eb-9a8c-82ac428c9533.png)
nie wiem jak tutaj wstrzyknąć token w takim wypadku

Reply:
To też nie dziala
![obraz](https://user-images.githubusercontent.com/19174792/94585924-f1290980-0280-11eb-800f-4d190a2b9816.png)

Reply:
Też nie mogę zrobić tak jak jest w 29 linijce, bo ta property jest readonly....
![obraz](https://user-images.githubusercontent.com/19174792/94587186-7fea5600-0282-11eb-926a-fd014337cb78.png)

Reply:
@michasacuer

Reply:
To się kompiluje
![obraz](https://user-images.githubusercontent.com/19174792/94591049-d3ab6e00-0287-11eb-9b7a-564e5d724f96.png)

tylko HttpClientInitializer jest nullem....
pomożesz z rejestracją tego w startupie?
na razie mam to (w Startupie.cs):
![obraz](https://user-images.githubusercontent.com/19174792/94591156-f5a4f080-0287-11eb-83ea-5ef59ab87663.png)

Reply:
TU już robi się syf, napisałem na fb. Wystaw drugiego merga najlepiej